docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+The solution is depicted in this diagram:
+
+![alt text](assets/CurrentClusterVPN.png "Overview Current Cluster VPN")
+
+These are the essential parts of the OpenVPN client configuration in the `vpn-seed` sidecar container:
+```
+# use TCP instead of UDP (commonly not supported by load balancers)
+proto tcp-client
+
+[...]
+
+# get all routing information from server
+pull
+
+tls-client
+key "/srv/secrets/vpn-seed/tls.key"
+cert "/srv/secrets/vpn-seed/tls.crt"
+ca "/srv/secrets/vpn-seed/ca.crt"
+
+tls-auth "/srv/secrets/tlsauth/vpn.tlsauth" 1
+cipher AES-256-CBC
+
+# https://openvpn.net/index.php/open-source/documentation/howto.html#mitm
+remote-cert-tls server
+
+# pull filter
+pull-filter accept "route 100.64.0.0 255.248.0.0"
+pull-filter accept "route 100.96.0.0 255.224.0.0"
+pull-filter accept "route 10.1.60.0 255.255.252.0"
+pull-filter accept "route 192.168.123."
+pull-filter ignore "route"
+pull-filter ignore redirect-gateway
+pull-filter ignore route-ipv6
+pull-filter ignore redirect-gateway-ipv6
+```
+
+Encryption is based on SSL certificates with an additional HMAC signature to all SSL/TLS handshake packets. As multiple clients connect to the OpenVPN server in the shoot cluster, all clients must be assigned a unique IP address. This is done by the OpenVPN server pushing that configuration to the client (keyword `pull`). As this is potentially problematic because the OpenVPN server runs in an untrusted environment there are pull filters denying all but IP address allocation and route distribution.
+
+The OpenVPN server running in the shoot cluster is configured as follows:
+
+```
+mode server
+tls-server
+proto tcp4-server
+dev tun0
+
+[...]
+
+server 192.168.123.0 255.255.255.0
+
+push "route 10.243.0.0 255.255.128.0"
+push "route 10.243.128.0 255.255.128.0"
+
+duplicate-cn
+
+key "/srv/secrets/vpn-shoot/tls.key"
+cert "/srv/secrets/vpn-shoot/tls.crt"
+ca "/srv/secrets/vpn-shoot/ca.crt"
+dh "/srv/secrets/dh/dh2048.pem"
+
+tls-auth "/srv/secrets/tlsauth/vpn.tlsauth" 0
+push "route 10.242.0.0 255.255.0.0"
+```
+
+It is a TCP TLS server and configured to automatically assign IP addresses for OpenVPN clients (`server` directive). In addition, it pushes the shoot cluster node-, pod-, and service ranges to the clients running in the seed cluster (`push` directive). 
+
+**Note:** The network mesh spanned by OpenVPN uses the network range `192.168.123.0 - 192.168.123.255`. This network range cannot be used in either shoot-, or seed clusters. If it is used this might cause subtle problem due to network range overlaps. Unfortunately, this appears not to be well documented but this restriction exists since the very beginning. We should clean up this technical debt as part of the excercise.
+
+### Goals
+
+* We intend to supersede the current VPN solution with the solution outlined in this proposal.
+* We intend to remove the code for the konnectivity tunnel once this solution proposal has been validated.
```

Comment:
What does "once this proposal has been validated" mean? If the feature gate is promoted to beta, for example?

Reply:
This proposal and the konnectivity tunnel address the same issue. I think it does not make sense to maintain both solutions in parallel forever, but we should decide at a point in time to go for one of them and remove the other one from the code base to make it leaner and easier to maintain.
That being said, the proposal for sure needs to run on a bigger number of clusters before I would see it as validated. As I am rather new to the gardener processes, I am unsure whether this means beta or GA of the feature gate.

Reply:
Yes, sure. OTOH, I'm wondering why we don't remove the `KonnectivityTunnel` feature gate right away in favor of this new `ReversedVPN`. The `KonnectivityTunnel` is alpha, not recommended for and not fully working in production, so we don't need to keep any "API"/backwards-compatibility promises here and could eliminate it even earlier?
In this proposal, you decide to go with a completely different solution, and you don't plan to invest into the `apiserver-network-proxy` anymore anyways.

Reply:
Fine with me.

Reply:
I would consider removal of the `KonnectivityTunnel` not to be part of this GEP and implementation PR #3812. We can however agree here that the `KonnectivityTunnel` can be removed once this solution has been validated.

Reply:
Which canary landscape/the open source project does not know about concrete landscapes?
I'm wondering why you propose keeping `KonnectivityTunnel` and all the code (which indeed causes a lot of complexities, branches, etc.) if we - with this GEP - clearly state that we don't continue pursuing it. Even if the new `ReversedVPN` feature can be implemented independently (is this true?), it probably makes sense to work on removal of `KonnectivityTunnel` (in a separate PR)?

Reply:
`ReversedVPN` is complete independent of the `KonnectivityTunnel`. Removal of the `KonnectivityTunnel` to me is therefore a different activity/PR.

Reply:
Yes, that's agreed already. Please proceed with `ReversedVPN` independent of the `KonnectivityTunnel`.

I'm just trying to get a picture of the plans with `KonnectivityTunnel` now that you all have invested significant time looking into the "next generation VPN". IIUC, you don't plan to pursue the `apiserver-network-proxy`-based solution (and it makes little sense to have two implementations for the same problem if there is no strong requirement for it). For me, this means that `KonnectivityTunnel` can be dropped entirely and without deprecation periods once `ReversedVPN` is implemented and merged to `master`, correct?

Nobody will maintain `KonnectivityTunnel` anymore, it's still alpha and very unstable, and it already causes some complexities in the code. I assume, `ReversedVPN` will only get merged when it's at least as good as `KonnectivityTunnel` (the expectation is that it'll work better, of course).
So, please present good/strong reasons for NOT removing `KonnectivityTunnel` if you think we should wait (what for?). 

/cc @mvladev

Reply:
From our point of view this should not be a blocker for this GEP, therefore resolving.

Reply:
OK. I understand your comment as agreement for the removal of KonnectivityTunnel as soon as ReversedVPN has been merged.